text/0000-prelude-2021-traits.md

```diff
@@ -0,0 +1,125 @@
+[minor][minor-change], such that some backwards compatibility breaks are potentially allowed, the possibility of relatively widespread breakage makes it less attractive to do so without an edition opt-in.
+
+The expected outcome is to make the newly added traits more accessible to users, potentially stimulating their use compared to alternatives.
+
+[minor-change]: https://rust-lang.github.io/rfcs/1105-api-evolution.html#minor-change-adding-a-defaulted-item
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The compiler currently brings all items from `std::prelude::v1` into scope for each module (for code using `std`). Crates that declare usage of the 2021 edition will instead get items from `std::prelude::edition2021` imported into scope. With the implementation of this RFC, we will add `edition2015` and `edition2018` modules in `std::prelude`, and a crate configuring a particular edition will get the prelude appropriate for its edition. The `v1` module will stick around for compatibility reasons, although it might be deprecated in the future.
+
+The `core` prelude will parallel the `std` prelude, containing the same structure and the same items as far as they are available in `core`.
+
+This RFC proposes to add the following traits to `std::prelude::edition2021`, but not `std::prelude::v1`:
+
+- `TryFrom`/`TryInto`
+- `FromIterator`
+
+Except for newly added traits, the `v1` and `edition2021` preludes should be kept in sync. Items that are added to the `edition2021` prelude should also be added to the `v1` prelude if there is negligible risk for compatibility. On the other hand, newer edition preludes may decide not to include items from an older edition's prelude.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The new prelude will be named after the edition that introduces it to make the link to the edition more obvious. Each edition will have a separate prelude, whether the contents are actually different or not.
+
+Importantly, we'll need to make the upgrade path to an edition containing new traits in the prelude as smooth as possible. The analysis can be done as follows:
+
+* Find method calls matching the name of one of the newly added traits' methods.
+  This can be done either by hardcoding these method names or by setting up some
+  kind of registry through the use of an attribute on the relevant traits.
+* After method resolution, if the method matches one of the newly added methods'
+  names but the originating trait is not from `core` or `std` (and/or does not
+  match the originating trait), suggest a rewrite to disambiguating syntax
+  (such as `foo.try_into()` to `TryInto::try_into(foo)`).
```

Comment:
How about auto(de)referencing? This might need to be `TryInto::try_into(&foo)` or `TryInto::try_into(&**foo)`, etc. in some cases.

Reply:
Hmm, maybe. To what level of detail should we go here? Presumably the method resolution will have figured out a concrete type to dispatch to, so we can figure out the amount of (de)referencing needed to get from the type of `foo` to the receiver type?

Reply:
Added a note to the migration lint algorithm code. Let me know what you think.